data/pigui/modules/new-game-window/recovery.lua

```diff
@@ -0,0 +1,43 @@
+-- Copyright © 2008-2024 Pioneer Developers. See AUTHORS.txt for details
+-- Licensed under the terms of the GPL v3. See licenses/GPL-3.txt
+
+local Json = require 'Json'
+local Lang = require 'Lang'
+local lui = Lang.GetResource("ui-core")
+
+local Recovery = {}
+
+-- patch
+--    from - number (version)
+--    to - number (version)
+--    apply(gameDoc) -> errorString? - function that modifies whole gameDoc
+--
+--    it is assumed that the patches are arranged in the array in ascending
+--    order of versions, and do not intersect
+local Patches = {}
+
+-- no patches yes
+
+-- the function returns the full game save document patched to the supported
+-- version or a string with an error
+function Recovery.loadDoc(saveName)
+	local gameDoc = Json.LoadSaveFile(saveName)
```

Comment:
Note: sooner or later, this function will be required to post-process the save game and fix up all contained table references. The order and location a table value *definition* is serialized in is non-deterministic and any JSON value encoding a reference to a specific table value may potentially be the one in which that table is serialized to.

As a result, I'd recommend this function go through all `{ ref: number }` values serialized in the `lua_modules_json` (and eventually `lua_modules_persistent`) tables and perform a partial unserialization / unpickling, enough to ensure the savefile is traversable as a "normal" Lua table with the appropriate table identities preserved.

Reply:
I did the substitution of refs, but did not do the unpickling - all readers unpickle what they need themselves, is it worth redoing? It seems that the reader code will not be simplified by this.
Although, after resolving the refs, I can simplify the equipment reader - there was a search using refs, which is now not needed.

Reply:
Did a complete unpickling before calling the readers. Now only the fixupDoc function in the Recovery module is responsible for compatibility with the Lua serializer.

Reply:
Creating readers is now less complicated and the pickling mechanism is now just an implementation detail - thanks for tackling that!